### DISCUSSION TITLE
Allow a dumper by oid rather than Python type

### DISCUSSION THREAD

**Discussion by pgjones** [psycopg/psycopg#666] [Ideas]
I'd like to do something like `conn.adapters.register_dumper(oid, CustomDumper)` such that psycopg always used the `CustomDumper` regardless of the Python type for the given postgres `oid`. I'd imagine it would be lower in precedence than a dumper for the actual type. I think this can be added to the Transformer code (if a dumper is not found by type).

**Comment by dvarrazzo** [comment]
Hello,

What is the use case for choosing dumper by oid?

The adapters map has a [`get_dumper_by_oid()`](https://www.psycopg.org/psycopg3/docs/api/adapt.html#psycopg.adapt.AdaptersMap.get_dumper_by_oid), which is used, for instance, by [`Copy.set_types()`](https://www.psycopg.org/psycopg3/docs/api/copy.html#psycopg.Copy.set_types). So, mechanisms for selecting dumpers by oid are already in the code.

How were yoy thinking to use them? In a normal query loop, how would you get the oid though?

**Reply by pgjones** [reply]
> What is the use case for choosing dumper by oid?

I'd like to support psycopg alongside existing support for asyncpg, which, as I understand, allows conversion based on the target type rather than argument type, see [docs](https://magicstack.github.io/asyncpg/current/usage.html#custom-type-conversions).

I did see the existing code, but I'm not sure how to adapt it to do what I'd like.

**Reply by dvarrazzo** [reply]
If we start from your sequence of parameters:

```python
(10, 'hello', datetime.date.today())
```

or your map of parameters:

```python
{"id": 10, "msg": 'hello', "date": datetime.date.today()}
```

how do you choose which OIDs to use?

**Reply by pgjones** [reply]
> how do you choose which OIDs to use?

I don't know - I hoped you'd know, or be better placed to understand how asyncpg works.

**Reply by dvarrazzo** [reply]
asyncpg has a different design, of which I'm not knowledgeable.

I've done some tests, to check if preparing a query returns type information, but it doesn't. Type information on a result are only available once a "portal" is created, which is after binding parameters to a prepared statement. So, in practice, it's the client which has to tell the server the types to use, the server won't suggest it.

So I don't see how to have the oid of the arguments specified automatically, if not looking at the Python types, and I doubt you want to call manually something like `cur.set_types(["int", "text", "date"])` before the query.

**Reply by pgjones** [reply]
Thanks for investigating. I'll get back if I figure something out about asyncpg, otherwise I'll look for a workaround.
